src/ast/ast.c

```diff
@@ -0,0 +1,328 @@
+            type == CYPHER_AST_MERGE  ||
+            type == CYPHER_AST_CREATE ||
+            type == CYPHER_AST_UNWIND
+            ) {
+            _AST_CollectAliases(&aliases, clause);
+        }
+    }
+
+    return aliases;
+}
+
+const cypher_astnode_t* AST_GetBody(const cypher_parse_result_t *result) {
+    const cypher_astnode_t *statement = cypher_parse_result_get_root(result, 0);
+    assert(statement && cypher_astnode_type(statement) == CYPHER_AST_STATEMENT);
+
+    return cypher_ast_statement_get_body(statement);
+}
+
+AST* AST_Build(cypher_parse_result_t *parse_result) {
+    AST *ast = rm_malloc(sizeof(AST));
+    ast->entity_map = NULL;
+    ast->root = AST_GetBody(parse_result);
+    assert(ast->root);
+    if (cypher_astnode_type(ast->root) == CYPHER_AST_QUERY) AST_BuildEntityMap(ast);
+
+    // Set thread-local AST
+    pthread_setspecific(_tlsASTKey, ast);
+
+    return ast;
+}
+
+AST* AST_NewSegment(AST *master_ast, uint start_offset, uint end_offset) {
+    AST *ast = rm_malloc(sizeof(AST));
+
+    uint n = end_offset - start_offset;
+
+    cypher_astnode_t *clauses[n];
+    for (uint i = 0; i < n; i ++) {
+        clauses[i] = (cypher_astnode_t*)cypher_ast_query_get_clause(master_ast->root, i + start_offset);
+    }
+    struct cypher_input_range range = {};
+    ast->root = cypher_ast_query(NULL, 0, (cypher_astnode_t *const *)clauses, n, NULL, 0, range);
+
+    pthread_setspecific(_tlsASTKey, ast); // TODO I don't know if I like this
+    AST_BuildEntityMap(ast);
+
+    return ast;
+}
+
+// TODO I find it so weird that this is necessary
+long AST_ParseIntegerNode(const cypher_astnode_t *int_node) {
+    assert(int_node);
+
+    const char *value_str = cypher_ast_integer_get_valuestr(int_node);
+    return strtol(value_str, NULL, 0);
+}
+
+/* TODO CYPHER_OP_IS_NULL and CYPHER_OP_IS_NOT_NULL are also unaries (though postfix).
+ * Investigate this, also check CIPs because I think null handling may be different in
+ * Cypher now than it is in the parser. */
+/*
+AST_Operator AST_ConvertUnaryOperatorNode(const cypher_astnode_t *expr) {
+    // TODO All in AR_EXP_FromExpression right now
+}
+*/
+
+bool AST_ClauseContainsAggregation(const cypher_astnode_t *clause) {
+    assert(clause);
+
+    bool aggregated = false;
+
+    // Retrieve all user-specified functions in clause.
+    TrieMap *referred_funcs = NewTrieMap();
+    AST_ReferredFunctions(clause, referred_funcs);
+
+    void *value;
+    tm_len_t len;
+    char *funcName;
+    TrieMapIterator *it = TrieMap_Iterate(referred_funcs, "", 0);
+    while(TrieMapIterator_Next(it, &funcName, &len, &value)) {
+        if(Agg_FuncExists(funcName)) {
+            aggregated = true;
+            break;
+        }
+    }
+    TrieMapIterator_Free(it);
+    TrieMap_Free(referred_funcs, TrieMap_NOP_CB);
+
+    return aggregated;
+}
+
+AST_Validation AST_PerformValidations(RedisModuleCtx *ctx, const AST *ast) {
+    char *reason;
+    AST_Validation res = AST_Validate(ast, &reason);
```

Comment:
Where does AST_Validate defined / implemented?

Reply:
`src/ast/ast_validations.c`

I felt that this separation was preferable, since both files are fairly long, but am open to other designs!

Reply:
👍 ast_validations.c seems like a good idea